### PR TITLE
[msan] Dynamically grow kNumStackOriginDescrs

### PR DESCRIPTION
**Pull request by thurstond**
StackOriginDescr/StackOriginPC are currently fixed size arrays
and have a FIXME of resizability. This patch dynamically doubles
the size of these arrays as needed (for runtime that is amortized linear
in the size of the array).

Since NumStackOriginDescrs (the number of items currently stored in the array)
is monotonically increasing, there is no need to ever decrease the
allocated size of the array.

Testing notes: the old fixed limit was 1 million variables; this would be an
extremely large test case. By manually capping the maximum size, I found
that param_tls_limit.cpp and compress_stack_depot.cpp are the tests that are most sensitive to
kNumStackOriginDescrs.

### CODE REVIEW

File: compiler-rt/lib/msan/msan.cpp [hunk 1]
```diff
@@ -312,6 +332,7 @@ static inline void SetAllocaOrigin(void *a, uptr size, u32 *id_ptr, char *descr,
   u32 id = *id_ptr;
   if (id == 0 || id == first_timer) {
     u32 idx = atomic_fetch_add(&NumStackOriginDescrs, 1, memory_order_relaxed);
+    ResizeStackOriginMetadata(idx);
```
Comment (kstoimenov) [inline compiler-rt/lib/msan/msan.cpp:335]: I am not sure if this is thread safe. I would recommend allocating StackOriginDescr to 1024 * 1024 in __msan_init and call ResizeStackOriginMetadata only to double the size. It still won't be 100% thread safe, but we expect it to happen very few times.

**Review by eugenis — CHANGES_REQUESTED** [review]
This is not thread safe. I do not think you can do realloc() if you want the mutation to be lock free.

We could simply increase the constant size of the buffer (not perfect, but it is just .bss, not real memory).

We could go with an array of pointers to exponentially increasing allocations (i.e. same as your case, but never deallocate, instead just allocate more and consider the container a logical concatenation of multiple buffers).

**Comment by thurstond** [comment]
> This is not thread safe. I do not think you can do realloc() if you want the mutation to be lock free.

Thanks for the review, I had overlooked thread safety.

> We could simply increase the constant size of the buffer (not perfect, but it is just .bss, not real memory).

This is implemented in https://github.com/llvm/llvm-project/pull/92838

> We could go with an array of pointers to exponentially increasing allocations (i.e. same as your case, but never deallocate, instead just allocate more and consider the container a logical concatenation of multiple buffers).

This is still tricky to implement in a thread-safe, lock-free manner, because there are two arrays to be grown (StackOriginDescrs, StackOriginPC).

**Comment by eugenis** [comment]
> > We could go with an array of pointers to exponentially increasing allocations (i.e. same as your case, but never deallocate, instead just allocate more and consider the container a logical concatenation of multiple buffers).
> 
> This is still tricky to implement in a thread-safe, lock-free manner, because there are two arrays to be grown (StackOriginDescrs, StackOriginPC).

I think it will actually work pretty well.

If subarrays are power of 2 sized (1, 2, 4, 8, 16, ...), then the subarray for index i is basically log2(i+1). Both arrays can be handled independently; once the index is acquired, the subarray pointer is initialized with a double-checked locking pattern.

**Comment by eugenis** [comment]
But then, it seems easier to just MAP_NORESERVE a large enough array in msan_init. It can be really large then, no concerns about relocation offset.

**Pull request closed without merging**